src/servers/src/prom.rs

```diff
@@ -463,6 +477,205 @@ pub async fn range_query(
+    matches: Matches,
+    db: Option<String>,
+}
+
+// Custom Deserialize method to support parsing repeated match[]
+impl<'de> Deserialize<'de> for Matches {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Matches, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        struct MatchesVisitor;
+
+        impl<'d> Visitor<'d> for MatchesVisitor {
+            type Value = Option<String>;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a string")
+            }
+
+            fn visit_map<M>(self, mut access: M) -> std::result::Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'d>,
+            {
+                let mut matches = String::new();
+                while let Some((key, value)) = access.next_entry::<String, String>()? {
+                    if key == "match[]" {
+                        matches.push_str(&value);
+                        // use $ as separator, because prometheus don't use $.
+                        matches.push('$');
+                    }
+                }
+                if matches.is_empty() {
+                    Ok(None)
+                } else {
+                    Ok(Some(matches))
+                }
+            }
+        }
+        Ok(Matches(deserializer.deserialize_map(MatchesVisitor)?))
+    }
+}
+
+#[axum_macros::debug_handler]
+pub async fn labels_query(
+    State(handler): State<PromHandlerRef>,
+    Query(params): Query<LabelsQuery>,
+    Form(form_params): Form<LabelsQuery>,
+) -> Json<PromJsonResponse> {
+    let matches: Option<Vec<String>> = params.matches.0.map(|s| {
+        s.split('$')
+            .filter(|s| !s.is_empty())
+            .map(|s| s.to_owned())
+            .collect()
+    });
+    let Some(queries) = matches.or_else(|| {
+        form_params.matches.0.map(|s| {
+            s.split('$')
+                .filter(|s| !s.is_empty())
+                .map(|s| s.to_owned())
+                .collect::<Vec<String>>()
+        })}
+    ) else { return PromJsonResponse::error("Unsupported", "match[] parameter is required"); };
+
+    let start = params
+        .start
+        .or(form_params.start)
+        .unwrap_or_else(day_before_current_time_rfc3339);
+    let end = params
+        .end
+        .or(form_params.end)
+        .unwrap_or_else(current_time_rfc3339);
+
+    let db = &params.db.unwrap_or(DEFAULT_SCHEMA_NAME.to_string());
+    let (catalog, schema) = super::parse_catalog_and_schema_from_client_database_name(db);
+    let query_ctx = Arc::new(QueryContext::with(catalog, schema));
+
+    let mut labels: HashSet<String> = HashSet::new();
+    labels.insert(METRIC_NAME.to_string());
+
+    for query in queries {
+        let prom_query = PromQuery {
+            query,
+            start: start.clone(),
+            end: end.clone(),
+            step: DEFAULT_LOOKBACK_STRING.to_string(),
```

Comment:
step and lookback are two different things. But we can use its value here. Better to leave a todo comment.